sycl/source/detail/device_filter.cpp

```diff
@@ -0,0 +1,116 @@
+//==------------------- device_filter.cpp ----------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/device_filter.hpp>
+#include <CL/sycl/info/info_desc.hpp>
+#include <detail/config.hpp>
+#include <detail/device_impl.hpp>
+
+#include <cstring>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+device_filter::device_filter(std::string &FilterString) {
+  const std::array<std::pair<std::string, info::device_type>, 5>
+      SyclDeviceTypeMap = {{{"host", info::device_type::host},
+                            {"cpu", info::device_type::cpu},
+                            {"gpu", info::device_type::gpu},
+                            {"acc", info::device_type::accelerator},
+                            {"*", info::device_type::all}}};
+  const std::array<std::pair<std::string, backend>, 4> SyclBeMap = {
+      {{"opencl", backend::opencl},
+       {"level_zero", backend::level_zero},
+       {"cuda", backend::cuda},
+       {"*", backend::all}}};
+
+  // handle the optional 1st entry, backend
+  size_t Cursor = 0;
+  size_t ColonPos = FilterString.find(":", Cursor);
+  auto It = std::find_if(
+      std::begin(SyclBeMap), std::end(SyclBeMap),
+      [=, &Cursor](const std::pair<std::string, backend> &Element) {
+        size_t Found = FilterString.find(Element.first, Cursor);
+        if (Found != std::string::npos) {
+          Cursor = Found;
+          return true;
+        }
+        return false;
+      });
+  if (It == SyclBeMap.end()) {
+    Backend = backend::all;
+  } else {
+    Backend = It->second;
+    if (ColonPos != std::string::npos) {
+      Cursor = ColonPos + 1;
+    } else {
+      Cursor = Cursor + It->first.size();
+    }
+  }
+
+  // handle the optional 2nd entry, device type
+  auto Iter = std::find_if(
+      std::begin(SyclDeviceTypeMap), std::end(SyclDeviceTypeMap),
+      [=, &Cursor](const std::pair<std::string, info::device_type> &Element) {
+        size_t Found = FilterString.find(Element.first, Cursor);
+        if (Found != std::string::npos) {
+          Cursor = Found;
+          return true;
+        }
+        return false;
+      });
+  if (Iter == SyclDeviceTypeMap.end()) {
+    DeviceType = info::device_type::all;
+  } else {
+    DeviceType = Iter->second;
+    ColonPos = FilterString.find(":", Cursor);
+    if (ColonPos != std::string::npos) {
+      Cursor = ColonPos + 1;
+    } else {
+      Cursor = Cursor + Iter->first.size();
+    }
+  }
+
+  // handle the optional 3rd entry, device number
+  if (Cursor < FilterString.size()) {
+    try {
+      DeviceNum = stoi(FilterString.substr(ColonPos + 1));
+      HasDeviceNum = true;
+    } catch (...) {
+      char message[100];
```

Comment:
Major comment: request for changes: 
Allocating 100 bytes seems unsafe because:
1) std::strcat() has undefined behavior when dest arg is not large enough to hold the result of dest+src.
2) Adding FilterString at the line 88 below potentially adds 1024 (accordingly to this - https://github.com/intel/llvm/pull/2239/files#diff-b8e9dd9b257e12a2d83fb6eef6e29764R19) or more symbols

Why not to use std::string here? I.e.:
std::string Message = std::string("Invalid device filter: ") + FilterString + "\nPossible backend values are {opencl,level_zero,cuda,*}." // and so on...

Reply:
Thanks. Done.